### ISSUE TITLE
Support of R(theta, phi) gates from QASM

### ISSUE THREAD

**Issue by yourball** [quantumlib/Cirq#3415]
Hi,

it looks like that R(theta, phi) gates are not supported by QASM parser in Cirq. 
It is especially strange having in mind that R(theta, phi) gate is hardware native gate for Google devices.

At the same time R gate is implemented in Cirq as cirq.PhasedXPowGate(p, t) and could be added in a circuit, but could not be parsed from QASM. Qiskit on the other hand supports QASM readout of the R gate.

I believe it shouldn't be very hard to add such functionality.

**Comment by balopat** [comment]
Thanks for opening this! 
Yes, this should be relatively easy to implement using a mapping like this: 

```
import cirq

from cirq.circuits.qasm_output import QasmUGate
import numpy as np

def r_gate(phi, theta):
    # based on the definition here:
    # https://github.com/Qiskit/qiskit-terra/blob/ca2bf14c5bc301c0656933a157ec545480bc0c55/qiskit/circuit/library/standard_gates/r.py#L62
    return QasmUGate(theta=theta, phi=phi - 0.5,
                     lmda=-phi + 0.5)

phi = 0.35
theta = 0.11
q = cirq.NamedQubit("a")
r = r_gate(phi, theta)(q)

# this is an equivalent mapping to p and t and s
xpow = cirq.PhasedXPowGate(phase_exponent=1 + phi, exponent=-theta, global_shift=-0.5)(q)

u_r = cirq.unitary(r)
u_x = cirq.unitary(xpow)
assert np.allclose(u_r, u_x)
```

**Comment by BillGatesNephew** [comment]
I was just looking at this issue, and if you're asking for the parser to create a `PhasedXPowGate` from QASM, then, based on my understanding, I don't think you can do it with the current version of the parser because it currently doesn't support parsing custom gate definitions. Using it with something like the following would throw an exception: 
```
OPENQASM 2.0;
include "qelib1.inc";
// R gate isn't in the above library so it must be defined, but the parser
// breaks here because it doesn't yet support custom gate definition syntax.
gate r(θ, φ) a {u3(θ, φ - π/2, -φ + π/2) a;}
// ... Other stuff ...
``` 
I guess you could add custom logic to check parameters for `u3` gates in `_parser.py` and return a `PhasedXPowGate` instead of a `QasmUGate`, but this would be tech debt that would likely need to be changed when the parser adds the ability to handle custom gate definitions down the road.

**Comment by dabacon** [comment]
This looks to me like it was fixed in #4455
